lib/spack/spack/database.py

```diff
@@ -109,24 +118,60 @@ def __init__(
             installed,
             ref_count=0,
             explicit=False,
-            installation_time=None
+            installation_time=None,
+            deprecated_for=None
     ):
         self.spec = spec
         self.path = str(path) if path else None
         self.installed = bool(installed)
         self.ref_count = ref_count
         self.explicit = explicit
         self.installation_time = installation_time or _now()
+        self.deprecated_for = deprecated_for
+
+    def install_type_matches(self, installed):
```

Comment:
The `installed` parameter is a boolean, and does not contain all of the information of an `install_status`. You have to combine it with `deprecated_for` to find out whether it is missing or deprecated.

I'll look at the idea of a `InstallStatuses.status_query()` method return an iterable of InstallStatus.

Reply:
If a user passes in `installed=False`, I think the behavior is equivalent to them passing in `installed=[InstallStatuses.DEPRECATED, InstallStatuses.MISSING]`. I was thinking that creating a method to convert these values might simplify the logic in this function. If you disagree then there's no need to make a change (as it is the logic is pretty well-contained).

Reply:
If they pass `installed=False`, that's the same as `installed=[InstallStatuses.MISSING]` because for backwards compatibility `installed=False` returns only missing (as it did before this PR).

I did create a method to convert these values, I think you're correct that it more clearly identifies what logic does what. It's updated now.